pkg/drivers/docker_driver.go

```diff
@@ -0,0 +1,371 @@
+	return env
+}
+
+// copies a tar archive starting at the specified path from the image, and returns
+// a tar reader which can be used to iterate through its contents and retrieve metadata
+func (d *DockerDriver) retrieveTar(path string) (*tar.Reader, error) {
+	// this contains a placeholder command which does not get run, since
+	// the client doesn't allow creating a container without a command.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image: d.currentImage,
+			Cmd:   []string{"NOOP_COMMAND_DO_NOT_RUN"},
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+	})
+	if err != nil {
+		logrus.Errorf("Error creating container: %s", err.Error())
+		return nil, err
+	}
+	defer d.removeContainer(container.ID)
+
+	var b bytes.Buffer
+	stream := bufio.NewWriter(&b)
+
+	err = d.cli.DownloadFromContainer(container.ID, docker.DownloadFromContainerOptions{
+		OutputStream: stream,
+		Path:         path,
+	})
+	if err != nil {
+		return nil, err
+	}
+	stream.Flush()
+	return tar.NewReader(bytes.NewReader(b.Bytes())), nil
+}
+
+func (d *DockerDriver) StatFile(target string) (os.FileInfo, error) {
+	reader, err := d.retrieveTar(target)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		switch header.Typeflag {
+		case tar.TypeDir, tar.TypeReg, tar.TypeLink, tar.TypeSymlink:
+			if filepath.Clean(header.Name) == path.Base(target) {
+				return header.FileInfo(), nil
+			}
+		default:
+			continue
+		}
+	}
+	return nil, fmt.Errorf("File %s not found in image", target)
+}
+
+func (d *DockerDriver) ReadFile(target string) ([]byte, error) {
+	reader, err := d.retrieveTar(target)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if filepath.Clean(header.Name) == path.Base(target) {
+				return nil, fmt.Errorf("Cannot read specified path: %s is a directory, not a file", target)
+			}
+		case tar.TypeSymlink:
+			return d.ReadFile(header.Linkname)
+		case tar.TypeReg, tar.TypeLink:
+			if filepath.Clean(header.Name) == path.Base(target) {
+				var b bytes.Buffer
+				stream := bufio.NewWriter(&b)
+				io.Copy(stream, reader)
+				return b.Bytes(), nil
+			}
+		default:
+			continue
+		}
+	}
+	return nil, fmt.Errorf("File %s not found in image", target)
+}
+
+func (d *DockerDriver) ReadDir(target string) ([]os.FileInfo, error) {
+	reader, err := d.retrieveTar(target)
+	if err != nil {
+		return nil, err
+	}
+	var infos []os.FileInfo
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		if header.Typeflag == tar.TypeDir {
+			// we only want top level dirs here, no recursion. to get these, remove
+			// trailing separator and split on separator. there should only be two parts.
+			parts := strings.Split(strings.TrimSuffix(header.Name, string(os.PathSeparator)), string(os.PathSeparator))
```

Comment:
Is this a sufficient check?
What does this actually look like?

Reply:
yep this works as intended. the idea here is that when we're reading a directory, we don't want to recursively dive into nested directories. so we strip off the tar header first, and then trim off the trailing separator (which gets appended to directories by the tar reader) and split the string on the separator. if we have two parts, we know we have a nested directory.

```
/foo